src/frequenz/sdk/timeseries/_formula_engine/_formula_printer.py

```diff
@@ -0,0 +1,220 @@
+        """Initialize the StackItem.
+
+        Args:
+            value: The value of the stack item.
+            precedence: The precedence of the stack item.
+        """
+        self.value = value
+        self.precedence = precedence
+
+    def __str__(self) -> str:
+        """Return the string representation of the stack item.
+
+        This is used for debugging purposes.
+
+        Returns:
+            str: The string representation of the stack item.
+        """
+        return f"(\"{self.value}\", {self.precedence})"
+
+    def plvalue(self, outer_precedence: OperatorPrecedence) -> str:
+        """Return the value of the stack item with parentheses if necessary.
+
+        Args:
+            outer_precedence: The precedence of the outer stack item.
+
+        Returns:
+            str: The value of the stack item with parentheses if necessary.
+        """
+        return f"({self.value})" if self.precedence < outer_precedence else self.value
+
+    def prvalue(self, outer_precedence: OperatorPrecedence) -> str:
+        """Return the value of the stack item with parentheses if necessary.
+
+        Args:
+            outer_precedence: The precedence of the outer stack item.
+
+        Returns:
+            str: The value of the stack item with parentheses if necessary.
+        """
+        return f"({self.value})" if self.precedence < outer_precedence else self.value
+
+    @staticmethod
+    def create_binary(lhs: StackItem, op: str, rhs: StackItem) -> StackItem:
+        """Create a binary stack item.
+
+        Args:
+            lhs: The left-hand side of the binary operation.
+            op: The operator of the binary operation.
+            rhs: The right-hand side of the binary operation.
+
+        Returns:
+            StackItem: The binary stack item.
+        """
+        pred = OperatorPrecedence.map(op)
+        return StackItem(
+            f"{lhs.plvalue(pred)} {op} {rhs.prvalue(pred)}",
+            pred,
+        )
+
+    @staticmethod
+    def create_primary(value: float) -> StackItem:
+        """Create a stack item for literal values or function calls (primary expressions).
+
+        Args:
+            value: The value of the literal.
+
+        Returns:
+            StackItem: The literal stack item.
+        """
+        return StackItem(str(value), OperatorPrecedence.Primary)
+
+
+class FormulaPrinter:
+    """Prints a formula."""
+
+    def __init__(self) -> None:
+        """Initialize the FormulaPrinter."""
+        self._stack = list[StackItem]()
+
+    @staticmethod
+    def print(postfix_expr: list[FormulaStep]) -> str:
+        """Return the formula as a string in infix notation.
+
+        Args:
+            postfix_expr: The steps of the formula in postfix notation order.
+
+        Returns:
+            str: The formula in infix notation.
+        """
+        printer = FormulaPrinter()
+        return printer._transform(postfix_expr)
```

Comment:
I wouldn't have a static method instantiating the class on each call for a couple of reasons:

1. There is no need to instantiate over and over again.
2. The class is more clumsy, as this is at the end just a plain function call, but you need to add some class name.

I think it would be more Pythonic to just make this a function and cache the instance in a global:

```python
_global_instance = FormulaFormatter()

def format(postfix_expr: list[FormulaStep]) -> str:
    return _global_instance.format(postfix_expr)
```

Reply:
We are having a member in that class (it might become more soon). I agree, I could just use a global here, but I am inherently against shared global state (!). I may as well just pass the members via call params, but that'll bloat up caller site.

I have renamed the (non-static) method to `_format` and and use this internally, which will also reduce the amount objects to be constructed.

Reply:
> We are having a member in that class (it might become more soon).

We can add more functions for then, if it makes sense.

> I agree, I could just use a global here, but I am inherently against shared global state (!). I may as well just pass the members via call params, but that'll bloat up caller site.

I think this is unjustified in Python and in the context of the SDK, which is not multi-threaded, we use globals in many cases, so just "I don't like shared global state" doesn't feel like a good justification against it by itself. Using the class name just as a namespace doesn't make any sense (and then the pointless extra instantiation too each time this is called).

Reply:
This is not part of the hot-path, people will print formulas once at the start of the application for example,  so there's no perf gain to be had from having a global variable, and it is not like in the other places where it was really necessary.

So I'm in favour of avoiding unnecessary globals as well.  And it is a class method now.